AK/HashTable.h

```diff
@@ -491,6 +496,141 @@ class HashTable {
         MUST(try_rehash(new_capacity));
     }
 
+    void rehash_in_place()
+    {
+        // FIXME: This implementation takes two loops over the entire bucket array, but avoids re-allocation.
+        //        Please benchmark your new implementation before you replace this.
+        //        The reason is that because of collisions, we use the special "rehashed" bucket state to mark already-rehashed used buckets.
+        //        Because we of course want to write into old used buckets, but already rehashed data shall not be touched.
+
+        // FIXME: Find a way to reduce the cognitive complexity of this function.
+
+        for (size_t i = 0; i < m_capacity; ++i) {
+            auto& bucket = m_buckets[i];
+
+            // FIXME: Bail out when we have handled every filled bucket.
+
+            if (bucket.state == BucketState::Rehashed || bucket.state == BucketState::End || bucket.state == BucketState::Free)
+                continue;
+            if (bucket.state == BucketState::Deleted) {
+                bucket.state = BucketState::Free;
+                continue;
+            }
+
+            auto const new_hash = TraitsForT::hash(*bucket.slot());
```

Comment:
It's unfortunate that we mod this in multiple places by capacity
Maybe spend 4-8 more bytes on the stack to save the scaled hash, to save on some divisions